drivers/adc/adc_ite_it8xxx2.c

```diff
@@ -0,0 +1,341 @@
+	CHIP_ADC_CH7,
+	CHIP_ADC_COUNT,
+};
+
+struct adc_it8xxx2_data {
+	struct adc_context ctx;
+	uint16_t *buffer;
+	uint16_t *repeat_buffer;
+};
+static void adc_it8xxx2_isr(const void *arg);
+
+static int adc_it8xxx2_channel_setup(const struct device *dev,
+				 const struct adc_channel_cfg *channel_cfg)
+{
+	ARG_UNUSED(dev);
+
+	if (channel_cfg->acquisition_time != ADC_ACQ_TIME_DEFAULT) {
+		LOG_ERR("Selected ADC acquisition time is not valid");
+		return -EINVAL;
+	}
+
+	if (channel_cfg->channel_id >= CHIP_ADC_COUNT) {
+		LOG_ERR("Channel %d is not valid", channel_cfg->channel_id);
+		return -EINVAL;
+	}
+
+	if (channel_cfg->gain != ADC_GAIN_1) {
+		LOG_ERR("Invalid channel gain");
+		return -EINVAL;
+	}
+
+	if (channel_cfg->reference != ADC_REF_INTERNAL) {
+		LOG_ERR("Invalid channel reference");
+		return -EINVAL;
+	}
+
+	if (channel_cfg->channel_id < CHIP_ADC_CH4) {
+		/*
+		 * for channel 0, 1, 2, and 3
+		 * bit4 ~ bit0 : indicates voltage channel[x]
+		 *               input is selected for measurement (enable)
+		 * bit5 : data valid interrupt of adc.
+		 * bit7 : W/C data valid flag
+		 */
+		*adc_ctrl_regs[channel_cfg->channel_id].adc_ctrl = 0xa0 +
+			channel_cfg->channel_id;
+	} else {
+		/*
+		 * for channel 4 ~ 7.
+		 * bit4 : voltage channel enable (ch 4~7)
+		 * bit5 : data valid interrupt of adc.
+		 * bit7 : W/C data valid flag
+		 */
+		*adc_ctrl_regs[channel_cfg->channel_id].adc_ctrl = 0xb0;
+	}
+
+	LOG_DBG("Channel setup succeeded!");
+
+	return 0;
+}
+
+static void adc_disable_channel(const struct device *dev, uint8_t ch)
+{
+	if (ch < CHIP_ADC_CH4) {
+		/*
+		 * for channel 0, 1, 2, and 3
+		 * bit4 ~ bit0 : indicates voltage channel[x]
+		 *               input is selected for measurement (disable)
+		 * bit 7 : W/C data valid flag
+		 */
+		*adc_ctrl_regs[ch].adc_ctrl = 0x9F;
+	} else {
+		/*
+		 * for channel 4 ~ 7.
+		 * bit4 : voltage channel disable (ch 4~7)
+		 * bit7 : W/C data valid flag
+		 */
+		*adc_ctrl_regs[ch].adc_ctrl = 0x80;
+	}
+
+	/* disable adc interrupt */
+	irq_disable(DT_INST_IRQN(0));
+}
+
+static int check_buffer_size(const struct adc_sequence *sequence,
+			     uint8_t active_channels)
+{
+	size_t needed_buffer_size;
+
+	needed_buffer_size = active_channels * sizeof(uint16_t);
+
+	if (sequence->options) {
+		needed_buffer_size *= (1 + sequence->options->extra_samplings);
+	}
+
+	if (sequence->buffer_size < needed_buffer_size) {
+		LOG_ERR("Provided buffer is too small (%u/%u)",
+				sequence->buffer_size, needed_buffer_size);
+		return -ENOMEM;
+	}
+
+	return 0;
+}
+
+static int adc_it8xxx2_start_read(const struct device *dev,
+			      const struct adc_sequence *sequence)
+{
+	struct adc_it8xxx2_data *data = DEV_DATA(dev);
+	int err;
+
+	LOG_DBG("Configure resolution=%d", sequence->resolution);
+
+	if (!sequence->channels ||
+		sequence->channels & ~BIT_MASK(CHIP_ADC_COUNT)) {
+		LOG_ERR("Invalid selection of channels");
+		return -EINVAL;
+	}
+
+	err = check_buffer_size(sequence, 1);
```

Comment:
This checks if the buffer is big enough for samples from one channel, but multiple channels can be selected.

Reply:
Add channel_count.